### ISSUE TITLE
Pinned / Locked tabs and tab groups like in Safari/Chrome

### ISSUE THREAD

**Issue by dawdmaow** [warpdotdev/Warp#2314]
### Describe the solution you'd like?

```text
I would like if Warp supported pinning tabs and moving them into tab groups basically the same way they function in Safari.
```

### How important is this feature to you?

2

### Warp Internal (ignore) - linear-label:770f6576-d6c0-4e4f-a259-fc64b5156087

None

**Comment by dannyneira** [comment]
Thanks for this feature request! Please :+1:  and subscribe to this request as it helps us gauge interest. We will post any updates here.

**Comment by yeliex** [comment]
`Edge` performs better than Chrome. the grouped tabs would reopen next start, and others closed

**Comment by bigloudjeff** [comment]
The tab grouping would be nice. A higher priority for me is tab pinning. Since sessions do not persist across windows being closed I would like to pin a tab in a way that prevents me from accidentally closing it.

**Comment by TwoAbove** [comment]
Bumping this up too, would be super convenient if CMD+D in pinned/protected terminals would be caught and ignored

**Comment by tobijafischer** [comment]
+1 for tab groups!
Being able to quickly open all the necessary terminal sessions associated with a particular project would make frequent project switching easier. With each tab having a persistent path, name, color and history.
Being able to quickly re-open all project related sessions would probably also prevent me from keeping all tabs open until it gets chaotic.

**Comment by Canopix** [comment]
+1 for Tabs Groups.

**Comment by byron1st** [comment]
+1 for pinned tab. Super cool!

**Comment by tokiory** [comment]
+1, it's a killer feature, when you have many projects opened in Warp!

**Comment by Neyunse** [comment]
+1

**Comment by danvayn** [comment]
Can't believe this hasn't been added yet.

**Comment by david-gettins** [comment]
This would help prevent accidental tab closes.

**Comment by ricpan** [comment]
This is a must have. Thank you!

**Comment by booleans-oss** [comment]
+1 Tab Groups

**Comment by dannyneira** [comment]
Hi Folks, to prevent spam, please avoid comments to show support. Instead add a 👍 to the original post at the top to signal that you want this feature, and subscribe if you'd like to be notified.

**Comment by ravingupta** [comment]
+1 for tab groups

**Comment by huix9527** [comment]
+1, sometimes close tabs by accident

**Comment by exymax** [comment]
+1
This is absolutely necessary for a product where engineers spend significant amount of their time, especially when working on multiple projects

**Comment by AlbertSu123** [comment]
Another good use case is when you want to leave something like [happy.engineering](https://happy.engineering/) and you don't want the tab to get closed accidentally since it's supposed to be always open

**Comment by elliottash** [comment]
+1 for pinned tabs, would be great.

**Comment by vainu-jussi** [comment]
+1

I'm closing tabs by accident also.

**Comment by bagusrizkis** [comment]
+1

**Comment by relliv** [comment]
+1

**Comment by danielquintas8** [comment]
Related discussion in #9063 — proposes a named Workspace layer above tabs (modeled on cmux's Workspace → Pane → Surface hierarchy). Likely the right consolidation point for the tab-groups half of this request.

**Comment by kasperhartwich** [comment]
@danielquintas8  Maybe the new task should be linked to this instead, instead of this task which is 3-4 years old.

**Comment by rakshitgondwal** [comment]
+1

**Comment by dil-atelegdi** [comment]
I'd add a +1 for both tab groups and pinned tabs! Very useful when working around in multiple projects simultaneously!

**Comment by captainsafia** [comment]
Our design team is gonna take a look at ironing out some mockups for this, particularly in the context of all vertical tabs work that has happened in previous releases.

**Comment by yaiceltg** [comment]
+1

**Comment by DevVali** [comment]
Yes, please! Warp is really changing the way we approach terminals as developers.

**Comment by sebet** [comment]
+1

**Comment by ap1392** [comment]
+1

**Comment by pythonzz0622** [comment]
+1

**Comment by nattfodd** [comment]
Tabs grouping is such a missable feature needed really daily. Workaround is to paint tabs, but its slow and not very effecient.

**Comment by apoapostolov** [comment]
New specific proposal filed as #12589: "Pinned tabs that remain in the vertical/horizontal tab line when closed (quick one-click reopen, retain custom name + color)".

This focuses on the 'closable but persistently visible in the tab chrome' UX (distinct from pure lock-to-prevent-close), so that customized tabs (name, color, etc.) survive intentional close and live as compact quick-launch entries directly in the (vertical or horizontal) tab bar.

See the detailed description there. Fits well with the vertical tabs + design work mentioned recently.

**Comment by feli0x** [comment]
bump +1

**Comment by dmit4mit** [comment]
+1

**Comment by Ale1x** [comment]
+1

**Comment by ap1392** [comment]
This issue can be closed since it was recently added in no? https://docs.warp.dev/changelog/2026/#20260703-v0202607010921
